test/tools/yulInterpreter/Interpreter.cpp

```diff
@@ -193,29 +197,41 @@ void Interpreter::operator()(Block const& _block)
 			break;
 	}
 
-	closeScope();
+	leaveScope();
 }
 
 u256 Interpreter::evaluate(Expression const& _expression)
 {
-	ExpressionEvaluator ev(m_state, m_dialect, m_variables, m_scopes);
+	ExpressionEvaluator ev(m_state, m_dialect, *m_scope, m_variables);
 	ev.visit(_expression);
 	return ev.value();
 }
 
 vector<u256> Interpreter::evaluateMulti(Expression const& _expression)
 {
-	ExpressionEvaluator ev(m_state, m_dialect, m_variables, m_scopes);
+	ExpressionEvaluator ev(m_state, m_dialect, *m_scope, m_variables);
 	ev.visit(_expression);
 	return ev.values();
 }
 
-void Interpreter::closeScope()
+void Interpreter::enterScope(Block const& _block)
+{
+	if (!m_scope->subScopes.count(&_block))
```

Comment:
What happens with recursive calls?

Reply:
This is just a syntactic tree, but I'll add a test!

Reply:
Actually there are already tests with recursive functions. The scope is just about visibility of variables and functions - the values of variables are tracked in another map that is freshly created for each function call.

Reply:
Ok!